docs/proposals/20230424-unify-the-definition-of-functions.md

```diff
@@ -0,0 +1,123 @@
+# Unify the definition of sync and async functions
+
+## Motivation
+Currently, we use `runtime: knative` and `runtime: async` to distinguish sync and async functions which increases the learning curve. Actually the difference between sync and async functions is trigger type:
+- Sync functions are triggered by `HTTP` event, which are defined by specifying `runtime: knative`
+- Async functions are triggered by events from components of `Dapr bindings` or `Dapr pubsub`. `runtime: async` and `inputs` have to be used together to specify triggers for async functions
+
+So we can use `triggers` to replace `runtime` and `inputs`.
+
+## Goals
+Use `triggers` to unify the definition of sync and async functions.
+
+## Proposal
+
+### OpenFunction
+
+We can use `triggers` to replace the original `inputs` field. The original `runtime` field will be replaced by `trigger.type`
+The API version can be upgrade from v1beta1 to v1
+
+```yaml=
+apiVersion: core.openfunction.io/v1beta1
+kind: Function
+metadata:
+  name: logs-async-handler
+spec:
+  version: "v2.0.0"
+  image: openfunctiondev/logs-async-handler:v1
+  imageCredentials:
+    name: push-secret
+  build:
+    builder: openfunction/builder-go:latest
+    env:
+      FUNC_NAME: "LogsHandler"
+      FUNC_CLEAR_SOURCE: "true"
+#     # Use FUNC_GOPROXY to set the goproxy if failed to fetch go modules
+#     FUNC_GOPROXY: "https://goproxy.cn"
+    srcRepo:
+      url: "https://github.com/OpenFunction/samples.git"
+      sourceSubPath: "functions/async/logs-handler-function/"
+      revision: "main"
+  serving:
+    ## The triggers definition is identical or similar to the origial inputs
+    ## The type of sync functions is http, http.knative or http.keda. http is identical to http.knative
+    ## The triggers.name and triggers.components is not used for sync functions
+    ## User can only specify one trigger if the trigger type is http/http.knative/http.keda
+    ## Supported trigger types for async functions are the same as Dapr pubsub and binding such as pubsub.rabbitmq and binding.kafka
+    ## Multiple triggers can be specified together if they're non-http triggers
+    triggers:
```

Comment:
how about having a `trigger` CRD as well? In this case, user can define trigger and reuse them:
```
triggers:
      - triggerRef: kafka
```

Reply:
> how about having a `trigger` CRD as well? In this case, user can define trigger and reuse them:
> 
> ```
> triggers:
>       - triggerRef: kafka
> ```

We already have a trigger crd in events

Reply:
I agree with the design of a CRD for trigger, but I think it could be considered together with function outputs, for example, like dapr Component, using a single CRD to characterize the inputs and outputs of the function

Reply:
> > how about having a `trigger` CRD as well? In this case, user can define trigger and reuse them:
> > ```
> > triggers:
> >       - triggerRef: kafka
> > ```
> 
> We already have a trigger crd in events

I think the Trigger CRD in the events framework is not a high priority and can be replaced by Core api if needed...

Reply:
> I agree with the design of a CRD for trigger, but I think it could be considered together with function outputs, for example, like dapr Component, using a single CRD to characterize the inputs and outputs of the function

The inputs and outputs are now defined with the same `DaprIO` type actually:
https://github.com/OpenFunction/OpenFunction/blob/main/apis/core/v1beta1/function_types.go#L184-L187

Reply:
> how about having a `trigger` CRD as well (or enhance in the future)? In this case, user can define triggers and reuse them:
> 
> ```
> triggers:
>       - triggerRef: kafka
> ```

The Trigger(Previously Inputs) is actually a name ref to the already defined dapr binding/pubsub components, we may consider add a dapr component template CR to avoid defining the dapr binding/pubsub components manually each time ( https://github.com/OpenFunction/samples/blob/main/functions/async/logs-handler-function/logs-handler-function.yaml#L61-L73 ). But it's not the scope of this proposal.